ssc/cmod_pvsamv1.cpp

```diff
@@ -2188,30 +2190,54 @@ void cm_pvsamv1::exec()
+            // AC connected batteries will set this laster
+            if (en_batt && (batt_topology == ChargeController::DC_CONNECTED)) {
+                batt->outGenWithoutBattery[idx] -= fabs(batt->outGenWithoutBattery[idx]) * PVSystem->acLossPercent * 0.01;;
+            }
 
             // Apply transformer loss
             ssc_number_t transformerRatingkW = static_cast<ssc_number_t>(PVSystem->ratedACOutput * util::watt_to_kilowatt);
             ssc_number_t xfmr_ll = PVSystem->transformerLoadLossFraction / step_per_hour;
             ssc_number_t xfmr_nll = PVSystem->transformerNoLoadLossFraction * static_cast<ssc_number_t>(ts_hour * transformerRatingkW);
 
-            if (PVSystem->transformerLoadLossFraction != 0 && transformerRatingkW != 0)
-            {
-                if (PVSystem->p_systemACPower[idx] < transformerRatingkW)
-                    xfmr_ll *= PVSystem->p_systemACPower[idx] * PVSystem->p_systemACPower[idx] / transformerRatingkW;
-                else
-                    xfmr_ll *= PVSystem->p_systemACPower[idx];
-            }
-            // total load loss
-            ssc_number_t xfmr_loss = xfmr_ll + xfmr_nll; // kWh
-            PVSystem->p_systemACPower[idx] -= xfmr_loss / ts_hour; // kW
+			if (PVSystem->transformerLoadLossFraction != 0 && transformerRatingkW != 0)
+			{
+				if (PVSystem->p_systemACPower[idx] < transformerRatingkW)
+					xfmr_ll *= PVSystem->p_systemACPower[idx] * PVSystem->p_systemACPower[idx] / transformerRatingkW;
+				else
+					xfmr_ll *= PVSystem->p_systemACPower[idx];
+			}
+			// total load loss
+			ssc_number_t xfmr_loss = xfmr_ll + xfmr_nll; // kWh
+			PVSystem->p_systemACPower[idx] -= xfmr_loss/ts_hour; // kW
+            if (en_batt && (batt_topology == ChargeController::DC_CONNECTED)) {
+                // Recompute transformer loss as if the battery didn't run
+                ssc_number_t xfmr_ll_no_batt = PVSystem->transformerLoadLossFraction / step_per_hour;
+                ssc_number_t xfmr_nll_no_batt = PVSystem->transformerNoLoadLossFraction * static_cast<ssc_number_t>(ts_hour * transformerRatingkW);
 
-            // transmission loss if AC power is produced
-            if (PVSystem->p_systemACPower[idx] > 0) {
-                PVSystem->p_systemACPower[idx] -= (ssc_number_t)(transmissionloss);
+                if (PVSystem->transformerLoadLossFraction != 0 && transformerRatingkW != 0)
+                {
+                    if (PVSystem->p_systemACPower[idx] < transformerRatingkW)
+                        xfmr_ll_no_batt *= batt->outGenWithoutBattery[idx] * batt->outGenWithoutBattery[idx] / transformerRatingkW;
+                    else
+                        xfmr_ll_no_batt *= batt->outGenWithoutBattery[idx];
+                }
+                // total load loss
+                ssc_number_t xfmr_loss_no_batt = xfmr_ll + xfmr_nll; // kWh
+                batt->outGenWithoutBattery[idx] -= xfmr_loss_no_batt / ts_hour;
             }
```

Comment:
Looks like this code was duplicated from earlier on in the code where it was used for `xfmr_ll`. Could you make a separate fx for calculating transformer losses?

Reply:
A function is tricky because of lines 2242 - 2255. With a function, annual_xfmr_nll, annual_xfmr_ll, and annual_xfmr_loss now have to become member variables of cm_pvsamv1, and there needs to be a boolean to tell the function whether or not to write to them (is this the actual loss, or the theoretical loss without the battery?) 

I can do the function if you prefer, but was hesitant to introduce the member variables as a part of this change.

Reply:
I agree with not adding unnecessary member variables. 

Would something like the below work? 

```
double transformerLoss(powerkW, transformerRatingkW, LoadLossFraction, NoLoadLossFraction, ts_hour, &xfmr_ll, $xfmr_nll) {
   if transformerRatingkW == 0 return 0

// calculate xfmr_ll, xfmr_nll
   
     if (powerkW < transformerRatingkW)
                        xfmr_ll *= powerKW * powerKW / transformerRatingkW;
                    else
                        xfmr_ll *= powerKW;
                }
                // total load loss
                return xfmr_ll + xfmr_nll; // kWh
}
```

Is line 2228 supposed to be `ssc_number_t xfmr_loss_no_batt = xfmr_ll_no_batt + xfmr_nll_no_batt;` ?

Reply:
Good catch on line 2228! I implemented your function suggestion in the latest push, thanks!

Reply:
Great! I think it's good to merge as long once Travis is done!